src/libraries/System.Text.Json/System.Text.Json.SourceGeneration/JsonSourceGenerator.cs

```diff
@@ -16,47 +18,105 @@ namespace System.Text.Json.SourceGeneration
     [Generator]
     public class JsonSerializerSourceGenerator : ISourceGenerator
     {
+        public Dictionary<string, Type> foundTypes = new Dictionary<string, Type>();
+
         public void Execute(SourceGeneratorContext context)
         {
-            // Foreach type found, call code generator.
-            StringBuilder sourceBuilder = new StringBuilder(@"
+            if (!(context.SyntaxReceiver is JsonSerializableSyntaxReceiver receiver))
+                return;
+
+            MetadataLoadContext mlc = new MetadataLoadContext(context.Compilation);
+
+            SemanticModel semanticModel;
+            INamedTypeSymbol namedTypeSymbol;
+            ITypeSymbol typeSymbol;
+
+            // Map type name to type objects.
+            foreach (KeyValuePair<string, TypeDeclarationSyntax> entry in receiver.InternalClassTypeDict)
+            {
+                TypeDeclarationSyntax tds = entry.Value;
+                semanticModel = context.Compilation.GetSemanticModel(tds.SyntaxTree);
+                namedTypeSymbol = (INamedTypeSymbol)semanticModel.GetDeclaredSymbol(tds);
+                Type convertedType = new TypeWrapper(namedTypeSymbol, mlc);
+                foundTypes[entry.Key] = convertedType;
+            }
+            foreach (KeyValuePair<string, IdentifierNameSyntax> entry in receiver.ExternalClassTypeDict)
+            {
+                IdentifierNameSyntax ins = entry.Value;
+                semanticModel = context.Compilation.GetSemanticModel(ins.SyntaxTree);
+                typeSymbol = context.Compilation.GetSemanticModel(ins.SyntaxTree).GetTypeInfo(ins).ConvertedType;
+                Type convertedType = new TypeWrapper(typeSymbol, mlc);
+                foundTypes[entry.Key] = convertedType;
+            }
+
+            // Create sources for all found types.
+            foreach (KeyValuePair<string, Type> entry in foundTypes)
+            {
+                context.AddSource($"{entry.Key}ClassInfo", SourceText.From($@"
 using System;
-namespace HelloWorldGenerated
-{
-    public static class HelloWorld
-    {
-        public static string SayHello() 
-        {
-            return ""Hello"";
-");
 
-            sourceBuilder.Append(@"
-        }
-    }
-}");
+namespace HelloWorldGenerated
+{{
+    public class {entry.Key}ClassInfo
+    {{
+        public {entry.Key}ClassInfo() {{ }}
 
-            context.AddSource("helloWorldGenerated", SourceText.From(sourceBuilder.ToString(), Encoding.UTF8));
+        public string TestMethod()
+        {{
+            return ""{entry.Key}"";
+        }}
+    }}
```

Comment:
Although this code will probably be updated/removed soon, we should generate methods that tell us the name, visibility, getter/setter if applicable, and attributes for each member on the type. This is to be sure that we actually obtain all the information that we need to generate source for each type, and we have a solid base to build upon for the generation phase.

Reply:
updated, ptal.